### DISCUSSION TITLE
Cannot find module postcss-import

### DISCUSSION THREAD

**Discussion by mackross** [flavorjones/tailwindcss-ruby#41] [Q&A] [answered]
Using v3.4.17 I used the example postcss.config.js from the documentation in tailwindcss-rails.

```
// postcss.config.js
module.exports = {
  plugins: {
    'postcss-import': {},
    'tailwindcss/nesting': {},
    tailwindcss: {},
    autoprefixer: {},
  },
}
```

when I run `bundle exec tailwindcss ---input ./app/assets/stylesheets/application.tailwind.css ./app/assets/builds/tailwind.css -postcss ` in the config directory I get the following error:

```
["/Users/mackross/Development/newapp/.devbox/virtenv/ruby/gems/tailwindcss-ruby-3.4.17-arm64-darwin/exe/arm64-darwin/tailwindcss", "--input", "../app/assets/stylesheets/application.tailwind.css", "../app/assets/builds/tailwind.css", "--postcss"]
node:internal/modules/cjs/loader:933
  const err = new Error(message);
              ^

Error: Cannot find module 'postcss-import'
Require stack:
- /Users/mackross/Development/newapp/config/postcss.config.js
1) If you want to compile the package/file into executable, please pay attention to compilation warnings and specify a literal in 'require' call. 2) If you don't want to compile the package/file into executable and want to 'require' it from filesystem (likely plugin), specify an absolute path in 'require' call using process.cwd() or process.execPath.
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function._resolveFilename (pkg/prelude/bootstrap.js:1951:46)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at Module.require (pkg/prelude/bootstrap.js:1851:31)
    at Module.Object.assign.resolve [as require] (/snapshot/tailwindcss/standalone-cli/patch-require.js:15:28)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/Users/mackross/Development/newapp/config/postcss.config.js:1:1)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Module._compile (pkg/prelude/bootstrap.js:1890:32) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/mackross/Development/newapp/config/postcss.config.js'
  ],
  pkg: true
}
```

Based on the docs for the tailwind-cli and this gem this should work? Any idea where to start? Thanks

**Comment by flavorjones** [comment, marked as answer]
@mackross The readme in this project, and [in the tailwind docs](https://tailwindcss.com/docs/using-with-preprocessors#using-post-css-as-your-preprocessor) assumes you already have postcss and the plugins installed via npm, and so some of the installation details are elided.

When you see the error message "Cannot find module 'postcss-import'" my understanding is that this is telling you that tailwindcss cannot find that module in the installed npm packages.

**Reply by mackross** [reply]
I’m not using npm, just this gem 💎. What I love about it is my build system doesn’t need to install any JavaScript related stuff.  I guess I expected this gem to include the dependencies that are supported it its base arguments.

**Reply by flavorjones** [reply]
How are you using postcss without npm? I don't understand.

**Reply by flavorjones** [reply]
Just to be clear, in case there's a misunderstanding: the postcss feature described in tailwindcss-rails's README is a compatibility feature for people who are already using postcss and npm. The assumption is that if you need that compatibility, you know what you're doing and have everything you need already installed.

Conversely, if you don't need it, then don't create the `postcss.config.js` file or use the `--postcss` argument.

Does that make sense? If not, I need to know more about what's motivating you to try to use a postcss config when you also say you don't want to use any JS-related tooling. I'm really failing to understand something.

**Reply by mackross** [reply]
At the moment I’m not using post-css. I was looking at it because I wanted nested and imports to work with tailwind and it seemed like the only way to do it. I guess where I’m coming from is if you’ve chosen to not use npm for your JS/css (the rails default) but you want tailwind to support the postcss features, it would be cool to have them bundled in. If I have to install npm for my css compilation, I might as well use cssbundling-rails as the win from using tailwindcss-rails is far smaller. Installing npm just for css was something I was hoping to avoid as the rest of my build is #nobuild.

**Reply by flavorjones** [reply]
@mackross I think either you are misunderstanding something fundamental, or I am.

My understanding, which is very limited, is that PostCSS is a javascript toolchain; and it is not possible to use PostCSS without javascript. Therefore it seems to me like a nonsensical statement to say you want to use postcss but you do not want to use javascript.

Orthogonally, this gem has nothing to do with postcss or javascript. The README for this gem says, at the very top (the first line!):

> A self-contained tailwindcss executable, wrapped up in a ruby gem. That's it. Nothing else.

If the tailwindcss executable does it, this gem does it. If the tailwindcss executable does not do it, this gem does not do it.

**Comment by mackross** [comment]
I get what you're saying https://github.com/rails/tailwindcss-rails < it's this readme which led me to think it would work with postcss. It's quite likely if you're using tailwindcss-rails you're not using npm.

**Reply by flavorjones** [reply]
I feel like we're going in circles a bit. I'm going to close this.
